File: 59-spiral-matrix-ii/spiral-matrix-ii.java
class Solution {
    public int[][] generateMatrix(int n) {
        int[][] matrix = new int[n][n];
        int num = 1;
        int top = 0, left = 0;
        int bottom = n - 1, right = n - 1;
        while (num <= n * n) {
            int i;
            for (i = left; i <= right; i++)
                matrix[top][i] = num++;
            top++;
            for (i = top; i <= bottom; i++)
                matrix[i][right] = num++;
            right--;
            for (i = right; i >= left; i--)
                matrix[bottom][i] = num++;
            bottom--;
            for (i = bottom; i >= top; i--)
                matrix[i][left] = num++;
            left++;
        }
        return matrix;
    }
}

File: 59-spiral-matrix-ii/spiral-matrix-ii.go
func generateMatrix(n int) [][]int {
    matrix := make([][]int, n)
    for i := range matrix {
        matrix[i] = make([]int, n)
    }

    num, top, left := 1, 0, 0
    bottom, right := n-1, n-1
    for num <= n*n {
        for i := left; i <= right; i++ {
            matrix[top][i] = num
            num++
        }
        top++
        for i := top; i <= bottom; i++ {
            matrix[i][right] = num
            num++
        }
        right--
        for i := right; i >= left; i-- {
            matrix[bottom][i] = num
            num++
        }
        bottom--
        for i := bottom; i >= top; i-- {
            matrix[i][left] = num
            num++
        }
        left++
    }
    return matrix
}
